py-polars/src/lazyframe.rs

```diff
@@ -584,6 +586,30 @@ impl PyLazyFrame {
         Ok(())
     }
 
+    #[allow(clippy::too_many_arguments)]
+    #[cfg(all(feature = "streaming", feature = "json"))]
+    #[pyo3(signature = (path, json_format, maintain_order))]
+    fn sink_json(
+        &self,
+        py: Python,
+        path: PathBuf,
+        json_format: Option<Wrap<JsonFormat>>,
+        maintain_order: bool
+    ) -> PyResult<()> {
+        let options = JsonWriterOptions {
+            json_format: json_format.map(|c| c.0).unwrap_or(JsonFormat::default()),
+            maintain_order,
+        };
+
+        // if we don't allow threads and we have udfs trying to acquire the gil from different
+        // threads we deadlock.
+        py.allow_threads(|| {
```

Comment:
also this here is duplicated every sink.

Can we not do something like 1 generic sink method where you need to pass in an instance of a BatchedWriter?
That way we can much more easily add more types to sink?

WDYT?

Reply:
Also agree on this one, will try to refactor this code as well

Reply:
After looking at the code, I don't think we can do much on this part 😞 there are only 2-3 lines of duplicated code because of `py.allow_threads` call, it would possible to remove the duplicated code if pyo3 implements an `allow_threads` attribute so the execution of the whole function releases the GIL.
It would be nice to be able to pass a BatchedWriter instance but then we will be mixing API and task execution code. I think is a better design to keep a separation of concerns that enables the API and the task's code to evolve in their own way.

Reply:
I'll think about it a bit more tomorrow & come back if I would think of a certain way to handle it.
link for my own benefit: (Parallelism)[https://pyo3.rs/main/parallelism]

Reply:
Thanks for looking into it. Nothing came to mind when I started to ponder about deadlocks and gil and stuff :-)